src/status_im2/subs/home.cljs

```diff
@@ -1,38 +1,20 @@
 (ns status-im2.subs.home
   (:require [re-frame.core :as re-frame]))
 
-(def memo-home-items (atom nil))
+(def memo-chats-stack-items (atom nil))
```

Comment:
should be fixed now

Reply:
Moved caching to the view itself instead of subscription. But looks like this slightly increase render time
<details>
<summary>Caching in subscription render time</summary>

![old](https://user-images.githubusercontent.com/17097240/231885497-348d9cc6-4aa7-41b0-8542-16dc9861a650.png)

</details>

<details>
<summary>Caching in view render time</summary>

![new](https://user-images.githubusercontent.com/17097240/231885614-50c36f3f-deb5-4f44-8936-a04f6040399a.png)


</details>

Reply:
probably render is also not the right place to memoize, ideally we use the re-frame subscription mechanism that already uses a cache internally?

Reply:
For now keeping the caching in the subscription itself, due to performance. We can look more into this later on how to improve it.

Reply:
Hey @cammellos, @rasom. @Parveshdhull recently mentioned this problem to me. I know it's been a while already, but here are my thoughts/findings. No action to be taken, just thoughts. This is going to be a long comment 🤷🏼 

The code in `develop` is still basically the same as when this PR was merged, so I used it as a basis to make the analysis.

> we want to keep data unchanged so react doesn't change component when we leave screen

In `develop`, when I switch between screens (1:1 chats, AC, bottom tabs, etc), the subscription `:chats-stack-items` always re-runs because it depends on the signal `:view-id`. The cost of this subscription when there's a cache hit in `memo-chats-stack-items` is fairly small, but it's not zero.

In `develop`, if we ignore the manual caching mechanism performed with `memo-chats-stack-items`, then the subscription `:chats-stack-items` is essentially a `take` call.

```clojure
(re-frame/reg-sub
 :chats-stack-items
 :<- [:chats/home-list-chats]
 :<- [:view-id]
 :<- [:home-items-show-number]
 (fn [[chats view-id home-items-show-number]]
   (take home-items-show-number chats)))
```

And upon further inspection, we see that the view layer does more computation on the results of this subscription.

```clojure
;; in component home.view/chats
(let [unfiltered-items (rf/sub [:chats-stack-items])
      items            (filter-and-sort-items-by-tab selected-tab unfiltered-items)]
  ...)
```

Therefore, the whole extra computation in `filter-and-sort-items-by-tab` is not being cached by re-frame.

After I moved all the computation to the subscription (including the current tab argument, this is important), then when I switch between bottom tabs or other screens, re-frame doesn't re-run the subscription the component `home.view/chats` never re-renders.

After switching screens, we get a hit in re-frame's subscription cache, which shows that when leaving or entering the *Messages home screen* nothing is re-computed or re-rendered. Obviously, if the user changes between `Recent` or `Groups` tabs, then the subscription will be re-computed and the view re-rendered.

```clojure
(re-frame.core/dispatch [:navigate-to :my-profile])
(re-frame.subs/cache-lookup [:chats-stack-items :tab/recent])
;; => still returns cached chats which means nothing is re-rendered or sub re-computed.
```

The new subscription looks like this:

```clojure
(re-frame/reg-sub
 :chats-stack-items
 :<- [:chats/home-list-chats]
 :<- [:home-items-show-number]
 (fn [[chats home-items-show-number] [_ tab]]
   (let [k (if (= tab :tab/groups) :group-chat :chat-id)]
     (->> chats
          (take home-items-show-number)
          (filter k)
          (sort-by :timestamp >)))))
```

Here's the example video showing the new sub at play:

[device-2023-08-10-123339.webm](https://github.com/status-im/status-mobile/assets/46027/6255e41c-3ccb-42d2-98d0-156d76e38d9a)

I measured with re-frisk and the new code delivers the same level of performance. The differences I've seen in my comparisons are more like random fluctuations, which makes me think this manual caching mechanism is not needed.

The upsides of the new code are:

- The subscription doesn't re-run on screen changes throughout the app.
- It is more idiomatic re-frame (no mutation in subscription).
- It's simpler and the whole computation can be unit tested (because it's not split between sub & view).

Side note: I micro-benchmarked the new subscription and the most expensive part is sorting, and because the collection is rather small for the value of `home-items-show-number` (e.g. 20), using a transducer didn't help because there is a negligible cost to garbage collect intermediary results. If the data coming through was already ordered by timestamp we could consider skipping sorting?

But this sorting detail is in the range of micro optimizations. I would say optimizing the individual render item component `chat-list-item/chat-list-item` can bear more fruit. For example, it unfortunately creates subscriptions for every item of the flat list; it passes the whole chat instance to the `notification` component which makes Reagent's life more difficult to diff, and so on.

Also, as far as I know, re-frame doesn't offer any mechanism to control when & how its `query-cache` is evicted (see https://github.com/day8/re-frame/blob/master/src/re_frame/subs.cljc#L46). One technique that's worth trying is playing with `reg-sub-raw` and the `on-dispose` callback. Sounds like a hack, but it's interesting nonetheless.

Complete sidetracking, but I can't help but wish https://github.com/clojure/core.memoize was available for ClojureScript, because then we would be able to use less naive memoization techniques that are less likely to grow too much, e.g. LRU or TTL semantics (see https://github.com/clojure/core.memoize/blob/master/src/main/clojure/clojure/core/memoize.clj#L398). I'm sure this would give us new guns to shoot ourselves in the foot, but I used core.memoize in the past in certain problems and it really opens some good doors.

That's all! 😅 Cheers!